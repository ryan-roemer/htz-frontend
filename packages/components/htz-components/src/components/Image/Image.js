import React from 'react';
import PropTypes from 'prop-types';
import { createComponent, FelaTheme, } from 'react-fela';
import Observer from 'react-intersection-observer';
import { parseComponentProp, parseStyleProps, } from '@haaretz/htz-css-tools';
import { stylesPropType, } from '../../propTypes/stylesPropType';
import { attrsPropType, } from '../../propTypes/attrsPropType';
import { buildURLs, buildUrl, } from '../../utils/buildImgURLs';
import ImgSource from './elements/ImgSource';
import DefaultImage from '../DefaultImage/DefaultImage';
import setColor from '../../utils/setColor';

const ImgWrapperStyle = ({ bgc, height, theme, width, miscStyles, }) => ({
  height: '0',
  // prettier-ignore
  paddingBottom: `${(height / width) * 100}%`,
  position: 'relative',
  width: '100%',
  extend: [
    parseComponentProp('backgroundColor', bgc || [ 'image', 'bgc', ], theme.mq, setColor, theme.color),
    // Trump all other styles with those defined in `miscStyles`
    ...(miscStyles ? parseStyleProps(miscStyles, theme.mq, theme.type) : []),
  ],
});

const StyledImgWrapper = createComponent(ImgWrapperStyle);

/** The shape of an `<Image />` component's transforms options */
const ImgTransfromOptions = PropTypes.shape({
  /** The image's aspect ratio to use as base crop, default 'full' */
  aspect: PropTypes.string,
  /** An array of strings with misc flags to apply to the image url */
  flags: PropTypes.arrayOf(PropTypes.string),
  /**
   * The height, in pixels, of the image the generated url points to.
   * This describes the number of pixels in the file (not rendered size!).
   */
  height: PropTypes.string,
  /** Generate a progressive jpeg. default false */
  isProgressive: PropTypes.bool,
  /** The image quality, default 'auto' */
  quality: PropTypes.string,
  /** An array of strings with misc transforms to apply to the image url */
  transforms: PropTypes.arrayOf(PropTypes.string),
  /**
   * The width, in pixels, of the image the generated url points to.
   * This describes the number of pixels in the file (not rendered size!).
   */
  width: PropTypes.string.isRequired,
});

const imageOptionsType = PropTypes.shape({
  /**
   * A string describing the sizes (optional) and the rendered
   * width of the image (not the file!).
   */
  sizes: PropTypes.string,
  /**
   * transforms for the url used as the `src` and `srcset` attributes.
   * When `transforms` is an array, the first item will be used for
   * the `src` attribute, and then all items will be used in
   * constructing the `srcset` attribute.
   */
  transforms: PropTypes.oneOfType([ ImgTransfromOptions, PropTypes.arrayOf(ImgTransfromOptions), ])
    .isRequired,
});

Image.propTypes = {
  /**
   * An object of attributes to set on the DOM element.
   * Passed to the underlying react element
   */
  attrs: attrsPropType,
  /**
   * A background color for the wrapper around the image, which will be
   * visible until the image is loaded. Defaults to the color set in the
   * theme.
   */
  bgcolor: PropTypes.string,
  /** Image data from polopoly */
  data: PropTypes.shape({
    /** Image alt from polopoly */
    alt: PropTypes.string,
    /**
     * When the array contains multiple elements, each should be used
     * independently in creating sources for the picture element.
     */
    imgArray: PropTypes.arrayOf(
      PropTypes.shape({
        /** Image name from polopoly by the format 'image/imageName.imgType' */
        imgName: PropTypes.string.isRequired,
        /** Image version from polopoly */
        version: PropTypes.string,
        /** Holds the image aspects object */
        aspects: PropTypes.object,
      })
    ),
    /** Image id from polopoly */
    contentId: PropTypes.string.isRequired,
    /** The photographer credit. added to title in the title attribute */
    credit: PropTypes.string,
    /**
     * When present, the image should be rendered inside
     * a `<picture>` element to serve webp image type
     */
    isAnimatedGif: PropTypes.bool,
  }).isRequired,
  /** An image options for user transforms and sizes */
  imgOptions: imageOptionsType.isRequired,
  /**
   * Determines if the image will be surrounded by
   * a wrapper to prevent content jumping
   */
  hasWrapper: PropTypes.bool,
  /** Passing attributes role="presentation" and aria-hidden="true" if exists  */
  isPresentational: PropTypes.bool,
  /**
   * Determine if the component should be lazyloaded. Defaults to `false`.
   * If lazyloaded, indicates how many pixels before entering the screen
   * should the image be loaded.
   * For example, when `{lazyLoad: true}`, the image will be
   * lazyloaded as soon as it enters the screen. When `{lazyLoad: '400px'}`
   * the image will be lazyloaded 400px before entering the screen.
   * Strings should be in css length units.
   */
  lazyLoad: PropTypes.oneOfType([ PropTypes.bool, PropTypes.string, ]),
  /** is image from extenal source. (dont use claudinary) */
  isExternal: PropTypes.bool,
  /**
   * A special property holding miscellaneous CSS values that
   * trumps all default values. Processed by
   * [`parseStyleProps`](https://Haaretz.github.io/htz-frontend/htz-css-tools#parsestyleprops).
   * passes down only to the image wrapper.
   */
  miscStyles: stylesPropType,
};

Image.defaultProps = {
  attrs: null,
  bgcolor: null,
  hasWrapper: true,
  isPresentational: false,
  lazyLoad: false,
  isExternal: false,
  miscStyles: null,
};

function Image(props) {
  const { data, imgOptions, } = props;
  if (data == null) {
    const { transforms, } = imgOptions;
    const { aspect, width, height, } = Array.isArray(transforms) ? transforms[0] : transforms;
    return <DefaultImage transforms={{ aspect, width, height, }} />;
  }

  const { attrs, bgcolor, isPresentational, lazyLoad, miscStyles, hasWrapper, } = props;
  const { accessibility, alt, title, credit, isAnimatedGif, } = data || {};
  const { sizes, } = imgOptions || {};

  const isPicture = data.imgArray.length > 1;
  if (isPicture) {
    console.error(
      `The data structure of the "${data.contentId}" image is of a picture element, not an image.
Please use the "<Picture />" component`
    );

    return null;
  }

  if (isPresentational && (attrs && (!!attrs.role || !!attrs['aria-hidden']))) {
    console.warn(
      'When "isPresentational" prop value is true, "role" and "aria-hidden" are set automatically'
    );
  }
  const { height, width, } = getDimensions(props);
  const sources = getSources(props);
  const src = sources[0];
  const srcSet = sources[1];
  const webpSources = isAnimatedGif ? getSources(props, true) : undefined;

  const Sources = (
    <FelaTheme
      render={theme => (isAnimatedGif ? (
        <picture>
          <ImgSource
            {...(sizes ? { sizes, } : {})}
            tagName="source"
            type="image/webp"
            src={webpSources[0]}
            {...(srcSet ? { srcSet: webpSources[1], } : {})}
          />

          <ImgSource
            alt={alt || accessibility}
            src={src}
            hasWrapper={hasWrapper}
            {...(srcSet ? { srcSet, } : {})}
            {...(sizes ? { sizes, } : {})}
            title={`${title || ''}${title && credit ? ', ' : ''}${
              credit ? `${theme.creditPrefixI18n.imageCreditPrefix}: ${credit}` : ''
            }`}
            attrs={
                isPresentational
                  ? {
                    ...attrs,
                    role: 'presentation',
                    'aria-hidden': true,
                  }
                  : attrs
              }
          />
        </picture>
      ) : (
        <ImgSource
          alt={alt || accessibility}
          hasWrapper={hasWrapper}
          title={`${title || ''}${title && credit ? ', ' : ''}${
            credit ? `${theme.creditPrefixI18n.imageCreditPrefix}: ${credit}` : ''
          }`}
          src={src}
          {...(srcSet ? { srcSet, } : {})}
          {...(sizes ? { sizes, } : {})}
          attrs={
              isPresentational
                ? {
                  ...attrs,
                  role: 'presentation',
                  'aria-hidden': true,
                }
                : attrs
            }
        />
      ))
      }
    />
  );

  if (!lazyLoad) {
    return hasWrapper ? (
      <StyledImgWrapper width={width} height={height} bgc={bgcolor} miscStyles={miscStyles}>
        {Sources}
      </StyledImgWrapper>
    ) : (
      Sources
    );
  }

  return hasWrapper ? (
    <StyledImgWrapper width={width} height={height} bgc={bgcolor} miscStyles={miscStyles}>
      <Observer triggerOnce rootMargin="1000px">
        {inView => (inView ? Sources : null)}
      </Observer>
    </StyledImgWrapper>
  ) : (
    <Observer triggerOnce rootMargin="1000px">
      {inView => (inView ? Sources : null)}
    </Observer>
  );
}

export default Image;

// //////////////////////////////////////////////////////////////////////
//                          Helper Functions                          //
// //////////////////////////////////////////////////////////////////////
export const aspectRatios = {
  regular: { height: 3, width: 3.9, },
  headline: { height: 9, width: 15.48, },
  landscape: { height: 9, width: 20.79, },
  square: { height: 1, width: 1, },
  vertical: { height: 20, width: 17, },
  belgrade: { height: 1, width: 3.18, },
};

function getDimensions({ data: { imgArray, }, imgOptions: { transforms, }, }) {
  const { aspect, } = Array.isArray(transforms) ? transforms[0] : transforms;
  const { height, width, } = imgArray[0].aspects[aspect] || aspectRatios[aspect];

  return { height, width, };
}

function getSources(
  { data: { contentId, imgArray, isAnimatedGif, }, imgOptions: { transforms, }, },
  isWebP = false
) {
  const transformsArray = Array.isArray(transforms) ? transforms : [ transforms, ];
  const hasSrcset = transformsArray.length > 1;
  const imageNameFromData = imgArray[0].imgName;

  const imgName = isAnimatedGif && isWebP ? `${imageNameFromData.split('.')[0]}.webp` : imageNameFromData;

  // Always use the first image in the imageArray
  const imgVersion = imgArray[0].version;
  const imgData = {
    imgName,
    version: imgVersion,
    aspects: imgArray[0].aspects,
  };

  const src = buildUrl(contentId, imgData, transformsArray[0]);

  return hasSrcset
    ? [
      // If more than one transform is defined,
      src, // Use the first transform to build the `src` attribute
      // And all transforms to build the `srcset` attribute
      buildURLs(contentId, imgData, transformsArray),
    ]
    : [ src, ];
}
