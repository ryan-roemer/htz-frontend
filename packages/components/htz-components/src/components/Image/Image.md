A component for rendering responsive images with 
built-in support for lazy-loading

### Features
* Img/source tag wrapped with responsive and styled div.
* Can render any type of image.
* Gif type will transform to webp type (for performance issues) and will be placed in a source tag.
* Comes handy when dealing with 1 image data or same image option aspects (e.g: not intended to treat info-graphic image)

#### **Basic use**

```jsx
<Image
  data={{
    isAnimatedGif: false,
    imgArray: [{ imgName: "image/2092337187.jpg", version: "1516524320" }],
    credit: "this is title/credit",
    alt: "this alt from data",
    contentId: "1.5748461",
    aspects: {
      full: { width: "1680", height: "1260", x: "0", y: "0" }
    }
  }}
  imgOptions={{
    transforms: { width: "1045", aspect: "full" }
  }}
  bgcolor={"black"}
/>
```

```jsx
<div>
  <Image
    data={{
      isAnimatedGif: false,
      imgArray: [{ imgName: "image/922278056.jpg", version: "1514197183" }],
      credit: "this is title/credit",
      caption: "this caption from data",
      alt: "this alt from data",
      aspects: {
        landscape: { width: "2220", height: "800", x: "10", y: "0" },
        full: { width: "500", height: "400", x: "0", y: "0" }
      },
      contentId: "1.4921360"
    }}
    imgOptions={{
      sizes: "(min-width:1420px) 610px,(min-width:1320px) 500px, 280px",
      transforms: [
        {
          width: "145",
          height: "100",
          aspect: "landscape",
          quality: "auto:best"
        },
        { width: "245", height: "159", aspect: "landscape" },
        { width: "445", aspect: "landscape" }
      ]
    }}
    lazyLoad={"500px"}
    isPresentational
    bgcolor={"black"}
  />
</div>
```

```jsx
<div>
  <Image
    data={{
      isAnimatedGif: true,
      imgArray: [{ imgName: "image/4030303706.gif", version: "1515097953" }],
      credit: "this is title/credit",
      caption: "this caption from data",
      alt: "this alt from data",
      aspects: {
        landscape: { width: "394", height: "200", x: "33", y: "0" },
        full: { width: "400", height: "400", x: "0", y: "0" }
      },
      contentId: "1.5599867"
    }}
    imgOptions={{
      sizes: "(min-width:1420px) 610px,(min-width:1320px) 500px, 280px",
      transforms: [
        {
          width: "145",
          height: "100",
          aspect: "landscape",
          quality: "auto:best"
        },
        { width: "245", height: "159", aspect: "landscape" },
        { width: "445", aspect: "landscape" }
      ]
    }}
    lazyLoad={"900px"}
    attr={{ "aria-hidden": true, role: "group" }}
  />
</div>
```

```jsx
<div>
  <Image
    data={{
      isAnimatedGif: false,
      imgArray: [
        { imgName: "image/918473823.jpg", version: "1516894981" },
        { imgName: "image/918473823.jpg", version: "1516894981" }
      ],
      credit: "טראמפ ונתניהו בדאבוס, היום - צילום: CARLOS BARRIA/רויטרס",
      alt: "טראמפ ונתניהו בדאבוס, היום",
      contentId: "1.5766455",
      aspects: {
        regular: { width: "2069", height: "1551", x: "59", y: "4" },
        landscape: { width: "2174", height: "1263", x: "0", y: "107" }
      }
    }}
    imgOptions={{
      transforms: [
        {
          width: "565",
          height: "320",
          aspect: "landscape",
          quality: "auto"
        },
        {
          width: "1040",
          height: "550",
          aspect: "regular",
          quality: "auto:best"
        }
      ]
    }}
    lazyLoad={"900px"}
  />
</div>
```
