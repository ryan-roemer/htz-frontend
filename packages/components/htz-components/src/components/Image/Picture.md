A component for rendering responsive images with
built-in support for lazy-loading

### Features

* Img/source tag wrapped with responsive and styled div.
* Can render any type of images.
* Gif type will transform to webp type (for performance issues) and will be placed in a source tag.
* Comes handy when dealing with more than image data or same image option aspects (e.g: intended to treat info-graphic image)

#### **Basic use**

### Picture component

```jsx
<Picture
  lazyLoad={"600px"}
  defaultImg={{
    sourceOptions: {
      transforms: { width: "1045", aspect: "full" }
    },
    data: {
      isAnimatedGif: false,
      imgArray: [{ imgName: "image/2092337187.jpg", version: "1516524320" }],
      credit: "this is title/credit",
      alt: "this alt from data",
      contentId: "1.5748461",
      aspects: {
        full: { width: "1680", height: "1260", x: "0", y: "0" }
      }
    }
  }}
  sources={[
    {
      until: "l",
      sourceOptions: {
        transforms: [
          { width: "645", height: "309", aspect: "landscape" },
          { width: "945", aspect: "landscape" }
        ]
      },
      data: {
        isAnimatedGif: true,
        imgArray: [{ imgName: "image/4030303706.gif", version: "1515097953" }],
        credit: "this is title/credit",
        caption: "this caption from data",
        alt: "this alt from data",
        aspects: {
          landscape: { width: "394", height: "200", x: "33", y: "0" },
          full: { width: "400", height: "400", x: "0", y: "0" }
        },
        contentId: "1.5599867"
      }
    },
    {
      from: "l",
      misc: "landscape",
      type: "screen",
      sourceOptions: { transforms: { width: "1045", aspect: "full" } },
      data: {
        isAnimatedGif: false,
        imgArray: [{ imgName: "image/2092337187.jpg", version: "1516524320" }],
        credit: "this is title/credit",
        alt: "this alt from data",
        contentId: "1.5748461",
        aspects: {
          full: { width: "1680", height: "1260", x: "0", y: "0" }
        }
      }
    }
  ]}
/>
```

```jsx
<Picture
  lazyLoad={"600px"}
  defaultImg={{
    sourceOptions: {
      transforms: [
        {
          width: "745",
          aspect: "headline",
          quality: "auto:best"
        },
        { width: "945", aspect: "headline" }
      ]
    },
    data: {
      isAnimatedGif: false,
      imgArray: [
        {
          imgName: "image/2196292809.jpg",
          version: "1517913716"
        }
      ],
      alt: "סיכות של קמפיים #METOO",
      credit: "LUCY NICHOLSON/רויטרס",
      contentId: "1.5791650",
      aspects: {
        headline: {
          width: "2048",
          height: "1192",
          x: "0",
          y: "74"
        }
      }
    }
  }}
  sources={[
    {
      until: "m",
      sourceOptions: { transforms: { width: "1045", aspect: "landscape" } },
      data: {
        isAnimatedGif: false,
        imgArray: [
          {
            imgName: "image/2196292809.jpg",
            version: "1517913716"
          }
        ],
        alt: "סיכות של קמפיים #METOO",
        credit: "LUCY NICHOLSON/רויטרס",
        contentId: "1.5791650",
        aspects: {
          landscape: {
            width: "2048",
            height: "883",
            x: "0",
            y: "70"
          }
        }
      }
    },
    {
      until: "xl",
      sourceOptions: {
        transforms: [
          {
            width: "745",
            aspect: "headline",
            quality: "auto:best"
          },
          { width: "945", aspect: "headline" },
          { width: "1215", aspect: "regular" }
        ]
      },
      data: {
        alt: "הכומר סת' קפר־דייל ורעייתו סטפני",
        credit: "נתן דביר",
        title:
          "הכומר סת' קפר־דייל ורעייתו סטפני. העולם' צריך להיות מקום מקלט אחד גדול'",
        aspects: {
          regular: {
            width: 1819,
            height: 1365,
            x: 96,
            y: 0
          },
          headline: {
            width: 2033,
            height: 1181,
            x: 15,
            y: 77
          }
        },
        isAnimated: false,
        imgArray: [
          {
            imgName: "image/392731171.jpg",
            version: "1518001678"
          }
        ],
        imageType: "image",
        inputTemplate: "com.tm.Image",
        contentId: "1.5803155",
        contentName: "Immigrants in Church"
      }
    },
    {
      from: "xl",
      misc: "landscape",
      mimeType: "image/webp",
      sourceOptions: {
        transforms: [
          {
            width: "345",
            aspect: "landscape",
            quality: "auto:best"
          },
          { width: "645", height: "309", aspect: "landscape" },
          { width: "945", aspect: "landscape" }
        ]
      },
      data: {
        alt: "איור של נרי יושבת על שפת מזרקה ולצידה צפרדע",
        title: "איתן אלוא",
        aspects: {
          landscape: {
            width: 1325,
            height: 569,
            x: "0",
            y: "90"
          }
        },
        isAnimated: false,
        imgArray: [
          {
            imgName: "image/3063850623.gif",
            version: "1518084212"
          }
        ],
        imageType: "image",
        inputTemplate: "com.tm.Image",
        contentId: "1.5804697",
        contentName: "איור נרי"
      }
    }
  ]}
/>
```
