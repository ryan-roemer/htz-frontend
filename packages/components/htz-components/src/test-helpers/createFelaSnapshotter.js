import React from 'react';
import renderer from 'react-test-renderer';
// eslint-disable-next-line import/no-extraneous-dependencies
import { renderToString, } from 'fela-tools';

// function Wrapper({children,}) {
//   return children;
// }
export default function createFelaSnapshotter(StyleProvider, testRenderer) {
  return function testFelaSnapshot(component) {
    const testObj = {
      // Output an object with both the component snapshot and raw styles.
      component: renderer
        .create(
          <StyleProvider renderer={testRenderer}>
            <div>{component}</div>
          </StyleProvider>
        )
        .toJSON(),
      styles: renderToString(testRenderer),
    };

    return testObj;
  };
}
