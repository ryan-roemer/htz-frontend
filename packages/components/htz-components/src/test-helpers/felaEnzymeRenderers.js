// eslint-disable-next-line import/no-extraneous-dependencies
import htzTheme from '@haaretz/htz-theme';
import { createRenderer, createFelaEnzymeRenderers, } from '@haaretz/fela-utils';
// import createFelaEnzymeRenderers from '../utils/createFelaEnzymeRenderers';

const testRenderer = createRenderer({ isRtl: true, });
const { felaMount, felaShallow, } = createFelaEnzymeRenderers(
  testRenderer,
  htzTheme
);

export { felaMount, felaShallow, };
