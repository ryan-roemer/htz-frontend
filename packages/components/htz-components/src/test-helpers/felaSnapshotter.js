import React from 'react';
import createRenderer from '../components/StyleProvider/createRenderer';
import StyleProvider from '../components/StyleProvider/StyleProvider';
import createFelaSnapshotter from './createFelaSnapshotter';

const testRenderer = createRenderer({ isRtl: true, });
const felaSnapshotter = createFelaSnapshotter(StyleProvider, testRenderer);

export default felaSnapshotter;
