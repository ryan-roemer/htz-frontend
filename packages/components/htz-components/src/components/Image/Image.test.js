import React from 'react';
import felaSnapshotter from '../../test-helpers/felaSnapshotter';
import { felaMount, felaShallow, } from '../../test-helpers/felaEnzymeRenderers';
import Image from './Image';

const mockData = {
  isAnimatedGif: false,
  imgArray: [
    {
      imgName: 'image/922278056.jpg',
      version: '1514197183',
    },
  ],
  credit: 'this is credit',
  alt: 'this alt from data',
  contentId: '1.4921360',
  aspects: {
    landscape: { width: '2220', height: '800', x: '10', y: '0', },
    full: { width: '500', height: '400', x: '0', y: '0', },
  },
};

describe('<Image />', () => {
  describe('DOM element', () => {
    it('Render correctly when essential props are passed ', () => {
      const { component, styles, } = felaSnapshotter(
        <Image
          data={mockData}
          imgOptions={{ transforms: { width: '1045', aspect: 'full', }, }}
        />
      );
      expect(component).toMatchSnapshot();
      expect(styles).toMatchSnapshot();
    });
    it('Render img correctly when all props are passed, lazyLoad=false ', () => {
      const { component, styles, } = felaSnapshotter(
        <Image
          data={mockData}
          imgOptions={{
            sizes: '(min-width:1420px) 610px,(min-width:1320px) 500px, 280px',
            transforms: [
              {
                width: '145',
                height: '100',
                aspect: 'landscape',
                quality: 'auto:best',
              },
              { width: '245', height: '159', aspect: 'landscape', },
              { width: '445', aspect: 'landscape', },
            ],
          }}
          lazyLoad={false}
          attrs={{ 'data-test': 'test', }}
          isPresentational
        />
      );
      expect(component).toMatchSnapshot();
      expect(styles).toMatchSnapshot();
    });
    it('Render picture correctly when all props are passed, lazyLoad=false ', () => {
      const { component, styles, } = felaSnapshotter(
        <Image
          data={{
            isAnimatedGif: true,
            imgArray: [
              { imgName: 'image/4030303706.gif', version: '1515097953', },
            ],
            credit: 'this is title/credit',
            caption: 'this caption from data',
            alt: 'this alt from data',
            aspects: {
              landscape: { width: '394', height: '200', x: '33', y: '0', },
              full: { width: '400', height: '400', x: '0', y: '0', },
            },
            contentId: '1.5599867',
          }}
          imgOptions={{
            sizes: '(min-width:1420px) 610px,(min-width:1320px) 500px, 240px',
            transforms: [
              {
                width: '145',
                height: '100',
                aspect: 'landscape',
                quality: 'auto:best',
              },
              { width: '245', height: '159', aspect: 'landscape', },
              { width: '445', aspect: 'landscape', },
            ],
          }}
          isPresentational
          attrs={{ border: '15px solid', }}
        />
      );
      expect(component).toMatchSnapshot();
      expect(styles).toMatchSnapshot();
    });

    it('Render picture tag when "isAnimatedGif" field is true value', () => {
      const wrapper = felaMount(
        <Image
          data={{
            isAnimatedGif: true,
            imgArray: [
              { imgName: 'image/4030303706.gif', version: '1515097953', },
            ],
            credit: 'this is title/credit',
            caption: 'this caption from data',
            alt: 'this alt from data',
            aspects: {
              landscape: { width: '394', height: '200', x: '33', y: '0', },
              full: { width: '400', height: '400', x: '0', y: '0', },
            },
            contentId: '1.5599867',
          }}
          imgOptions={{
            sizes: '(min-width:1420px) 610px,(min-width:1320px) 500px, 280px',
            transforms: [
              {
                width: '145',
                height: '100',
                aspect: 'landscape',
                quality: 'auto:best',
              },
              { width: '245', height: '159', aspect: 'landscape', },
              { width: '445', aspect: 'landscape', },
            ],
          }}
        />
      );
      expect(wrapper.find('picture').length).toBe(1);
      expect(wrapper.find('source').length).toBe(1);
      expect(wrapper.find('img').length).toBe(1);
    });
    it('Has attributes of role="presentation" and aria-hidden="true" when isPresentational prop is passed', () => {
      const wrapper = felaMount(
        <Image
          data={mockData}
          imgOptions={{
            sizes: '(min-width:1420px) 610px,(min-width:1320px) 500px, 280px',
            transforms: [
              {
                width: '145',
                height: '100',
                aspect: 'landscape',
                quality: 'auto:best',
              },
              { width: '245', height: '159', aspect: 'landscape', },
              { width: '445', aspect: 'landscape', },
            ],
          }}
          isPresentational
        />
      );
      expect(
        wrapper.find('img').find({ role: 'presentation', 'aria-hidden': true, })
          .length
      ).toBe(1);
    });

    it('Prints warning on console when both "isPresentational" and "attr" has keys of "role" or "aria-hidden" ', () => {
      const wrapper = felaShallow(
        <Image
          data={mockData}
          imgOptions={{
            sizes: '(min-width:1420px) 610px,(min-width:1320px) 500px, 280px',
            transforms: [
              {
                width: '145',
                height: '100',
                aspect: 'landscape',
                quality: 'auto:best',
              },
              { width: '245', height: '159', aspect: 'landscape', },
              { width: '445', aspect: 'landscape', },
            ],
          }}
          attrs={{ 'aria-hidden': true, role: 'group', }}
          isPresentational
        />
      );
      console.warn = jest.fn();
      global.console = { warn: jest.fn(), };
    });
  });
});
