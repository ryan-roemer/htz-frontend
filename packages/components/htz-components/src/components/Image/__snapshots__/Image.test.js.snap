// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> DOM element Render correctly when essential props are passed  1`] = `
<div>
  <div
    className="htz-a htz-b htz-c htz-d htz-e"
  >
    <img
      alt="this alt from data"
      className="htz-f htz-g htz-h htz-i htz-d"
      src="https://images.haarets.co.il/image/upload/w_500,h_400,x_0,y_0,c_crop,g_north_west/q_auto,h_836,w_1045,c_fill,f_auto/fl_any_format.preserve_transparency.progressive:none/v1514197183/1.4921360.922278056.jpg"
      title="this is credit"
    />
  </div>
</div>
`;

exports[`<Image /> DOM element Render correctly when essential props are passed  2`] = `".htz-a{height:0}.htz-b{padding-bottom:80%}.htz-c{position:relative}.htz-d{width:100%}.htz-e{background-color:#EBEBEB}.htz-f{height:100%}.htz-g{left:0}.htz-h{position:absolute}.htz-i{top:0}"`;

exports[`<Image /> DOM element Render img correctly when all props are passed, lazyLoad=false  1`] = `
<div>
  <div
    className="htz-a htz-b htz-c htz-d htz-e"
  >
    <img
      alt="this alt from data"
      aria-hidden={true}
      className="htz-f htz-g htz-h htz-i htz-d"
      data-test="test"
      role="presentation"
      sizes="(min-width:1420px) 610px,(min-width:1320px) 500px, 280px"
      src="https://images.haarets.co.il/image/upload/w_2220,h_800,x_10,y_0,c_crop,g_north_west/q_auto:best,h_100,w_145,c_fill,f_auto/fl_any_format.preserve_transparency.progressive:none/v1514197183/1.4921360.922278056.jpg"
      srcSet={
        Array [
          "https://images.haarets.co.il/image/upload/w_2220,h_800,x_10,y_0,c_crop,g_north_west/q_auto:best,h_100,w_145,c_fill,f_auto/fl_any_format.preserve_transparency.progressive:none/v1514197183/1.4921360.922278056.jpg 145w",
          "https://images.haarets.co.il/image/upload/w_2220,h_800,x_10,y_0,c_crop,g_north_west/q_auto,h_159,w_245,c_fill,f_auto/fl_any_format.preserve_transparency.progressive:none/v1514197183/1.4921360.922278056.jpg 245w",
          "https://images.haarets.co.il/image/upload/w_2220,h_800,x_10,y_0,c_crop,g_north_west/q_auto,h_160,w_445,c_fill,f_auto/fl_any_format.preserve_transparency.progressive:none/v1514197183/1.4921360.922278056.jpg 445w",
        ]
      }
      title="this is credit"
    />
  </div>
</div>
`;

exports[`<Image /> DOM element Render img correctly when all props are passed, lazyLoad=false  2`] = `".htz-a{height:0}.htz-b{padding-bottom:36.03603603603604%}.htz-c{position:relative}.htz-d{width:100%}.htz-e{background-color:#EBEBEB}.htz-f{height:100%}.htz-g{left:0}.htz-h{position:absolute}.htz-i{top:0}"`;

exports[`<Image /> DOM element Render picture correctly when all props are passed, lazyLoad=false  1`] = `
<div>
  <div
    className="htz-a htz-b htz-c htz-d htz-e"
  >
    <picture>
      <source
        sizes="(min-width:1420px) 610px,(min-width:1320px) 500px, 240px"
        srcSet={
          Array [
            "https://images.haarets.co.il/image/upload/w_394,h_200,x_33,y_0,c_crop,g_north_west/q_auto:best,h_100,w_145,c_fill,f_auto/fl_any_format.preserve_transparency.progressive:none/v1515097953/1.5599867.4030303706.webp 145w",
            "https://images.haarets.co.il/image/upload/w_394,h_200,x_33,y_0,c_crop,g_north_west/q_auto,h_159,w_245,c_fill,f_auto/fl_any_format.preserve_transparency.progressive:none/v1515097953/1.5599867.4030303706.webp 245w",
            "https://images.haarets.co.il/image/upload/w_394,h_200,x_33,y_0,c_crop,g_north_west/q_auto,h_226,w_445,c_fill,f_auto/fl_any_format.preserve_transparency.progressive:none/v1515097953/1.5599867.4030303706.webp 445w",
          ]
        }
        type="image/webp"
      />
      <img
        alt="this alt from data"
        aria-hidden={true}
        border="15px solid"
        className="htz-f htz-g htz-h htz-i htz-d"
        role="presentation"
        sizes="(min-width:1420px) 610px,(min-width:1320px) 500px, 240px"
        src="https://images.haarets.co.il/image/upload/w_394,h_200,x_33,y_0,c_crop,g_north_west/q_auto:best,h_100,w_145,c_fill,f_auto/fl_any_format.preserve_transparency.progressive:none/v1515097953/1.5599867.4030303706.gif"
        srcSet={
          Array [
            "https://images.haarets.co.il/image/upload/w_394,h_200,x_33,y_0,c_crop,g_north_west/q_auto:best,h_100,w_145,c_fill,f_auto/fl_any_format.preserve_transparency.progressive:none/v1515097953/1.5599867.4030303706.gif 145w",
            "https://images.haarets.co.il/image/upload/w_394,h_200,x_33,y_0,c_crop,g_north_west/q_auto,h_159,w_245,c_fill,f_auto/fl_any_format.preserve_transparency.progressive:none/v1515097953/1.5599867.4030303706.gif 245w",
            "https://images.haarets.co.il/image/upload/w_394,h_200,x_33,y_0,c_crop,g_north_west/q_auto,h_226,w_445,c_fill,f_auto/fl_any_format.preserve_transparency.progressive:none/v1515097953/1.5599867.4030303706.gif 445w",
          ]
        }
        title="this is title/credit"
      />
    </picture>
  </div>
</div>
`;

exports[`<Image /> DOM element Render picture correctly when all props are passed, lazyLoad=false  2`] = `".htz-a{height:0}.htz-b{padding-bottom:50.76142131979695%}.htz-c{position:relative}.htz-d{width:100%}.htz-e{background-color:#EBEBEB}.htz-f{height:100%}.htz-g{left:0}.htz-h{position:absolute}.htz-i{top:0}"`;
